docs/devguides/dependency_mgmt.md

```diff
@@ -0,0 +1,223 @@
+
+This is how we achieve that:
+
+1. When a component is rendered, it inserts an HTML comment containing metadata about the rendered component.
+
+   So a template like this
+
+   ```django
+   {% load component_tags %}
+   <div>
+     {% component "my_table" / %}
+   </div>
+   {% component "button" %}
+     Click me!
+   {% endcomponent %}
+   ```
+
+   May actually render:
+
+   ```html
+   <div>
+     <!-- _RENDERED "my_table_10bc2c,c020ad" -->
+     <table>
+       ...
+     </table>
+   </div>
+   <!-- _RENDERED "button_309dcf,31c0da" -->
+   <button>Click me!</button>
+   ```
+
+   Each `<!-- _RENDERED -->` comment includes comma-separated data - a unique hash for the component class, e.g. `my_table_10bc2c`, and the component ID, e.g. `c020ad`.
+
+   This way, we or the user can freely pass the rendered around or transform it, treating it as a string to add / remove / replace bits. As long as the `<!-- _RENDERED -->` comments remain in the rendered string, we will be able to deduce which JS and CSS dependencies the component needs.
+
+2. Post-process the rendered HTML, extracting the `<!-- _RENDERED -->` comments, and instead inserting the corresponding JS and CSS dependencies.
+
+   If we dealt only with JS, then we could get away with processing the `<!-- _RENDERED -->` comments on the client (browser). However, the CSS needs to be processed still on the server, so the browser receives CSS styles already inserted as `<style>` or `<link>` HTML tags. Because if we do not do that, we get a [flash of unstyled content](https://en.wikipedia.org/wiki/Flash_of_unstyled_content), as there will be a delay between when the HTML page loaded and when the CSS was fetched and loaded.
+
+   So, assuming that a user has already rendered their template, which still contains `<!-- _RENDERED -->` comments, we need to extract and process these comments.
+
+   There's multiple ways to achieve this:
+
+   - The approach recommended to the users is to use the `ComponentDependencyMiddleware` middleware, which scans all outgoing HTML, and post-processes the `<!-- _RENDERED -->` comments.
+
+   - If users are using `Component.render()` or `Component.render_to_response()`, these post-process the `<!-- _RENDERED -->` comments by default.
+
+     - NOTE: Users are able to opt out of the post-processing by setting `kind=None`.
+
+   - For advanced use cases, users may use `render_dependencies()` directly. This is the function that both `ComponentDependencyMiddleware` and `Component.render()` call internally.
+
+   `render_dependencies()`, whether called directly, via middleware or other way, does the following:
+
+   1. Find all `<!-- _RENDERED -->` comments, and for each comment:
+
+   2. Look up the corresponding component class.
+
+   3. Get the component's inlined JS / CSS from `Component.js/css`, and linked JS / CSS from `Component.Media.js/css`.
+
+   4. Generate JS script that loads the JS / CSS dependencies.
```

Comment:
Short answer, no, because the CSS is also inserted as `<style>` and `<link>` tags.

But this raises a good question on whether components would still render correctly if the browser had disabled execution of javascript. [Playwright supports this option](https://stackoverflow.com/questions/65958243/disable-javascript-in-playwright), so we should be able to test that, so I'll do that in one of the follow-up MRs.

Reply:
To elaborate a bit more, currently there's the `kind="document"` mode, which renders the CSS as said above (inserted as `<style>` and `<link>` tags). So components with `kind="document"` SHOULD be renderable even in no-JS browsers.

However, this is not true for `kind="fragment"`. There, we do need JS to load the JS / CSS dependencies.

Reply:
Understood!